internal/sync/value/latest.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package value
+
+import "sync/atomic"
+
+// A value that can be atomically updated, where each update invalidates the
+// previous value. Whenever the value changes, an associated expiration channel
+// is closed. Callers can use this to be notified of updates. The zero value of
+// Latest holds a zero value of T.
+//
+// Latest is useful when some shared state is updated frequently, and readers
+// don't need to keep track of every value, just the latest one. Latest makes
+// this easy, as there's no need to maintain a separate channel for each reader.
+//
+// Example Usage:
+//
+//	package main
+//
+//	import (
+//		"fmt"
+//		"sync"
+//		"time"
+//
+//		"github.com/k0sproject/k0s/internal/sync/value"
+//	)
+//
+//	func main() {
+//		// Declare a zero latest value
+//		var l value.Latest[int]
+//
+//		fmt.Println("Zero value:", l.Get()) // Output: Zero value: 0
+//
+//		// Set the value
+//		l.Set(42)
+//		fmt.Println("Value set to 42")
+//
+//		// Peek at the current value and get the expiration channel
+//		value, expired := l.Peek()
+//		fmt.Println("Peeked value:", value) // Output: Peeked value: 42
+//
+//		// Use a goroutine to watch for expiration
+//		var wg sync.WaitGroup
+//		wg.Add(1)
+//		go func() {
+//			defer wg.Done()
+//			<-expired
+//			fmt.Println("Value expired, new value:", l.Get()) // Output: Value expired, new value: 84
+//		}()
+//
+//		// Set a new value, which will expire the previous value
+//		time.Sleep(1 * time.Second) // Simulate some delay
+//		l.Set(84)
+//		fmt.Println("New value set to 84")
+//
+//		wg.Wait() // Wait for the watcher goroutine to finish
+//	}
+type Latest[T any] struct {
```

Comment:
An alternative name could be `Observable`. The channel provides a way to "observe" the changes, right. 😄

Reply:
Yeah, but that's resembling too much the ReactiveX naming for my taste.

Reply:
Maybe we could use RxGo then 🙈

Reply:
Other ideas:
* `Now`
* `Currently`

Reply:
I'm also good with `Latest`, just wanted to throw out some suggestion. This is internal only stuff so naming is not _that_ important, we can change that later on too if needed/wanted.